public/app/percona/shared/components/hooks/recurringCall.hook.ts

```diff
@@ -0,0 +1,37 @@
+import { logger } from '@percona/platform-core';
+import { useRef, useEffect } from 'react';
+
+type hookRecurringCallback = () => void;
+
+export const useRecurringCall = () => {
+  const timer = useRef<number>();
+  const interval = useRef<number>();
+
+  const triggerTimeout = async (cb: hookRecurringCallback, defaultInterval = 10000, callImmediate = false) => {
+    interval.current = defaultInterval;
+    try {
+      callImmediate && (await cb());
+      timer.current = window.setTimeout(async () => {
+        await cb();
+        triggerTimeout(cb, interval.current);
+      }, interval.current);
+    } catch (e) {
+      logger.error(e);
+      triggerTimeout(cb, interval.current);
+    }
+  };
+
+  const changeInterval = (newInterval: number) => {
+    interval.current = newInterval;
+  };
+
+  const stopTimeout = () => {
+    window.clearTimeout(timer.current);
+  };
+
+  useEffect(() => {
+    return stopTimeout;
+  }, []);
+
+  return [triggerTimeout, changeInterval, stopTimeout] as const;
+};
```

Comment:
What's the benefit of using a hook for this? Looks like an added complication, as opposed to just using `setTimeout` and cancel timeout on unmount. This is against KISS, since it's hiding the implementation and it's just *apparently* simpler for the user.

Reply:
I get your point. My concern here was that this pattern will come up at every page on Backup Management. That means that, for every page:
- creating a var to hold reference to the timer ID;
- creating a function that recursively calls `getData()` or similar;
- cleaning up the timer on unmount;

VS:
```
const [triggerTimeout] = useRecurringCall();
triggerTimeout(getData, DATA_INTERVAL)
```

And the hook gives you ability to change the time between calls, which is something someone suggested, although it's probably not worth it now. Sure, that can be achieved as well from just using setTimeout from the outside, but then that's another variable you have to store.

It's simpler for the user and it is indeed hiding the implementation, that's the whole point of hooks I guess. If not, why then:
- `DBCluster.hook` -> used once, why encapsulate it away from the user?
- `parameters.hooks` -> why not call `getLocationSrv().update(...)` every single time instead?
- `pageTitle.hook` -> why not just doing `document.title = 'whatever'` every time as well?
- `PageWrapper` -> why not removing the `Breadcrumb` and add it as needed from the outside?

I believe this is pretty much in line with what we've doing, correct me if I'm wrong.

Reply:
We use the hook to solve replication problems and to make the component simpler by separating some of the logic (e.g. separate view/component from dealing with processing the return of the API call) so I agree with the use of the hook here, even though it has pros and cons. On the other hand I think it's too generic how it's being used here. All the possible use cases you mention are related to Backup Management so this could go inside the Backup Management feature. In `DBCluster.hooks` the hook is also processing the output of the call and it deals with multiple calls at once. What I mean by this is that each use case is different so I see no point in generalising it now and ending up with a hook in the shared folder only being used in one place. If we need it later we can always generalise. What do you think?

Reply:
Sure @tiagomotasantos, seems like a reasonable approach 👍

Reply:
Indeed @tiagomotasantos, that's what I mean! As we discussed several times, let's avoid this "we might use it again" and optimize when really needed, because we might also not need it later (YAGNI) :)

> It's simpler for the user and it is indeed hiding the implementation, that's the whole point of hooks I guess

Well, the point of hooks is to put side effects in isolated and reusable units, not to hide the implementation :) It simplifies complex components, but sometimes a solution might become more complex when isolated from a component (e.g.: abstract away just a `setTimeout` used in a `useEffect` in a component) than if left in the component itself. In such cases the reason behind creating hooks is to avoid duplication.

That's why if it's not solving duplication *now*, it's just increasing complexity. Again, please, let's avoid "premature refactoring", see https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)

Reply:
I get it. But as Tiago mentioned, this is probably more related to Backup Management, so it's ok if that's the only use for it.
What I'm saying is that I'll need to use this logic on (at least) Backup Inventory, Restore History and Scheduled Backups, so it's not exactly an "YAGNI" case or premature refactoring, I was trying to save myself from some predictably repeated logic on all those Backup Management pages, hence the hook.

Reply:
The YAGNI principle states exactly that: don't use something unless you really need it now :) See how Jeffries puts it: https://ronjeffries.com/xprog/articles/practices/pracnotneed/.

The problem is that in order to have a more generic solution, you spend more time 1) foreseeing that you'll need it and 2) adjusting to that, thus prematurely optimizing. While we should try to define the problem in its smallest scope possible and work on that, leaving the refactoring to a later stage (which doesn't mean we're introducing tech debt, on which we incur if we delay that refactoring). That helps who writes the code to write it faster and the reviewers to better understand it (more concise code, with less abstractions).

Obviously I'm not saying to change it now, but just to keep this in mind.

Reply:
Absolutely. Thanks for the feedback 👍